### ISSUE TITLE
Bump CMake minimum required version 

### ISSUE THREAD

**Issue by jcfr** [python-cmake-buildsystem/python-cmake-buildsystem#116]
To simplify the build system and remove some of the workaround needed to support older version of CMake, I was thinking to bump the minimum required version of CMake. 

Looking at the table below (compiled using detailed data reported further down), bumping to CMake 2.8.11 will allow to use the CMake executable provided by most of the system (expect Centos 5 and 7).

Or we could bump the requirement to CMake 3.5 and expect developer to either download the latest CMake binaries or build CMake from source.

What do you all think ?

| System | Available CMake version |
| --- | --- |
| Debian stable (Jessie) | 3.0.2 |
| Ubuntu 15.10 LTS (wily) | 3.2.2 |
| Fedora 23 | 3.4.1 |
| Manylinux (based of Centos 5.11) | 2.6.4 |
| CentOS 7 | 2.8.11 |
| CentOS 6 | 2.8.12 |
| CentOS 5 | 2.6.4 |
| OpenSuse 13.X | 3.2.2 |
| RedHat EL 5,6,7 | 2.8.11 |
| Windows (Chocolatey) | 3.4.3 |
| MacOSX (Homebrew) | 3.5.0 |
| MacOSX (MacPort) | 3.5.1 |

To help make a decision, as of March 29th 2016, I collected these info:
- [manylinux](https://github.com/pypa/manylinux/blob/master/docker/Dockerfile-x86_64) is based of `centos:5.11` and provides `cmake-2.6.4-5.el5.4.x86_64.rpm`. See http://mirror.centos.org/centos/5.11/os/x86_64/CentOS/
- chocolatey: provides CMake 3.4.3. See https://chocolatey.org/packages/cmake
- homebrew:  Formula for `cmake-3.5.0`. See https://github.com/Homebrew/homebrew/blob/master/Library/Formula/cmake.rb
- macport: `cmake-3.5.1`. See https://www.macports.org/ports.php?by=library&substr=cmake
- debian: https://packages.debian.org/search?keywords=cmake

```
    wheezy (oldstable) (devel): cross-platform, open-source make system
    2.8.9-1: amd64 armel armhf i386 ia64 kfreebsd-amd64 kfreebsd-i386 mips mipsel powerpc s390 s390x sparc
    wheezy-backports (devel): cross-platform, open-source make system
    2.8.11.1-1~bpo70+1: amd64 armel armhf i386 ia64 kfreebsd-amd64 kfreebsd-i386 mips mipsel powerpc s390 s390x sparc
    jessie (stable) (devel): cross-platform, open-source make system
    3.0.2-1: amd64 arm64 armel armhf i386 mips mipsel powerpc ppc64el s390x
    stretch (testing) (devel): cross-platform, open-source make system
    3.5.0-1: amd64 arm64 armel armhf i386 mips mipsel powerpc ppc64el s390x
    sid (unstable) (devel): cross-platform, open-source make system
    3.5.0-1: alpha amd64 arm64 armel armhf hppa hurd-i386 i386 m68k mips mipsel powerpc ppc64 ppc64el s390x sh4 sparc64 x32
    3.2.2-2+b1: kfreebsd-amd64 kfreebsd-i386
```
- Ubuntu: http://packages.ubuntu.com/search?searchon=names&keywords=cmake

```
    precise (12.04LTS) (devel): cross-platform, open-source make system
    2.8.7-0ubuntu4: amd64 i386
    precise-updates (devel): cross-platform, open-source make system
    2.8.7-0ubuntu5: amd64 i386
    trusty (14.04LTS) (devel): cross-platform, open-source make system
    2.8.12.2-0ubuntu3: amd64 i386
    vivid (devel): cross-platform, open-source make system
    3.0.2-1ubuntu2: amd64 i386
    wily (devel): cross-platform, open-source make system
    3.2.2-2ubuntu3: amd64 i386
    xenial (devel): cross-platform, open-source make system
    3.3.2is3.2.2-0ubuntu5: amd64 i386
```
- Gentoo: `CMake 3.5.1` See https://packages.gentoo.org/packages/dev-util/cmake
- RPM based distrib (https://rpm.pbone.net/):

```
Scientific Linux 7  ftp.sourceforge.net/pub/sourceforge/s/sl/sl7-i686-project/yum/FEDOREL7/FULLMISSING/cmake-2.8.11-4.el7.i686.rpm
Scientific Linux 6  ftp.scientificlinux.org/linux/scientific/6rolling/x86_64/os/Packages/cmake-2.8.12.2-4.el6.x86_64.rpm
Scientific Linux 5  ftp.scientificlinux.org/linux/scientific/5rolling/x86_64/SL/cmake-2.6.4-5.el5.4.x86_64.rpm
RedHat EL 7     pkgs.repoforge.org/cmake/cmake-2.8.8-1.el7.rfx.x86_64.rpm
RedHat EL 6     pkgs.repoforge.org/cmake/cmake-2.8.8-1.el6.rfx.x86_64.rpm
RedHat EL 5     pkgs.repoforge.org/cmake/cmake-2.8.8-1.el5.rfx.x86_64.rpm
OpenSuSE 13.X   packman.iu-bremen.de/suse/openSUSE_13.1/Essentials/x86_64/cmake-3.2.2-200.2.x86_64.rpm
OpenSuSE 12.X   packman.iu-bremen.de/suse/openSUSE_12.3/Essentials/x86_64/cmake-3.2.2-200.1.x86_64.rpm
OpenSuSE 11.X   packman.iu-bremen.de/suse/openSUSE_11.4/Essentials/x86_64/cmake-3.1.3-184.1.x86_64.rpm
Mageia 5    distrib-coffee.ipsl.jussieu.fr/mageia/distrib/5/x86_64/media/core/updates/cmake-3.0.2-1.mga5.x86_64.rpm
Mageia 4    distrib-coffee.ipsl.jussieu.fr/mageia/distrib/4/x86_64/media/core/release/cmake-2.8.12.1-1.mga4.x86_64.rpm
Mageia 3    distrib-coffee.ipsl.jussieu.fr/mageia/distrib/3/x86_64/media/core/release/cmake-2.8.10.2-5.mga3.x86_64.rpm
Fedora 23   download.fedora.redhat.com/pub/fedora/linux/updates/23/x86_64/c/cmake-3.4.1-4.fc23.x86_64.rpm
Fedora 22   download.fedora.redhat.com/pub/fedora/linux/updates/22/x86_64/c/cmake-3.3.2-1.fc22.x86_64.rpm
Fedora 21   ftp5.gwdg.de/pub/opensuse/repositories/devel:/cloverleaf:/testing/Fedora_21/x86_64/cmake-3.2.1-3.1.fc21.x86_64.rpm
Fedora 20   ftp5.gwdg.de/pub/opensuse/repositories/Openwsman/Fedora_20/x86_64/cmake-3.0.0-142.1.x86_64.rpm
Fedora 19   ftp5.gwdg.de/pub/opensuse/repositories/Openwsman/Fedora_19/x86_64/cmake-3.0.0-142.1.x86_64.rpm
Fedora 18   ftp5.gwdg.de/pub/opensuse/repositories/Openwsman/Fedora_18/x86_64/cmake-2.8.11.2-106.1.x86_64.rpm
Fedora 17   ftp5.gwdg.de/pub/opensuse/repositories/Openwsman/Fedora_17/x86_64/cmake-2.8.11.2-106.1.x86_64.rpm
CentOS 7    ftp.centos.org/7.2.1511/os/x86_64/Packages/cmake-2.8.11-5.el7.x86_64.rpm
CentOS 6    ftp.centos.org/6.7/os/x86_64/Packages/cmake-2.8.12.2-4.el6.x86_64.rpm
CentOS 5    ftp.centos.org/5.11/os/x86_64/CentOS/cmake-2.6.4-5.el5.4.x86_64.rpm
```

List above obtained by:
(1) manually copying output of search illustrated below into file `~/Projects/cmake-pbone-list.txt`
(2) running `cat ~/Projects/cmake-pbone-list.txt | sort -r | uniq > ~/Projects/cmake-pbone-list-filtered.txt`
(3) manually editing files `~/Projects/cmake-pbone-list-filtered.txt` keeping the first entry of each distrib

![2016 03 29_cmake_pbone_search](https://cloud.githubusercontent.com/assets/219043/14119459/3fb47926-f5bc-11e5-9ec7-134235a36347.png)

**Comment by jcfr** [comment]
Cc: @chuckatkins @python-cmake-buildsystem/team-python-cmake-buildsystem

**Comment by adrianbroher** [comment]
Maybe something else to consider:

travis-ci only provides 14.04 Ubuntu with their `trusty` image and a regular 12.04 Ubuntu image.  These in respect provide a 2.8.12 or 2.8.7 cmake packaging.

**Comment by thewtex** [comment]
Great analysis!

+1 for 2.8.11 so workarounds for old bugs and available features are reasonable. For older systems, we can get / provide CMake.

**Comment by chuckatkins** [comment]
I'd argue that it should be bumped at least to 3.X, which introduced a lot
of new features for targets carrying their dependencies and usage
requirements with them.  Since this package isn't looking to build a system
supplied python, we shouldn't necessarily be limited by the system supplied
cmake.  So if the version is going to be bumped anyways, it should be to
one that provides a meaningful new feature set.

**Comment by dbrnz** [comment]
This issue has been dormant for over three years... @jcfr, what are your thoughts on updating the minimum required version of CMake to a recent 3.x release?

Using CMake 3.14 results in a number of deprecation warnings regarding policies `CMP0042`, `CMP0045`, and `CMP0075`. I also note that `Extensions.cmake` contains commented out code marked as `Uncomment when CMake >= 2.8.8 is required`; are there other files that have been flagged in a similar way?

**Comment by jcfr** [comment]
You are right, it is about time to move on. If you want to start a branch, I would be happy to review and help move things forward.

**Comment by jcfr** [comment]
The following table was generated on Jan 3rd 2021 using link reported in the comment [above](https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/issues/116#issue-144355938) was well the following links:
* https://pkgs.org/download/cmake
* https://github.com/Homebrew/homebrew-cask/blob/master/Casks/cmake.rb
* https://en.wikipedia.org/wiki/Red_Hat_Enterprise_Linux#RHEL_9

| System | Available CMake version |
| --- | --- |
| Debian stable (Bulleye) | 3.18.4 |
| Ubuntu 18.04 LTS (Bionic) | 3.10 |
| Ubuntu 20.04 LTS (Focal) | 3.16 |
| Ubuntu 21.04 (Hirsute) | 3.18.4 |
| Ubuntu 21.10 (Impish) | 3.18.4 |
| Ubuntu 22.04 (Jammy) | 3.21.4 |
| Fedora 34 | 3.20.5 |
| Fedora 35 | 3.22.1 |
| Fedora 28 | 3.14.4 |
| Manylinux (based of Centos 5.11) | 2.6.4 |
| Manylinux2010 (based on Centos 6) | 2.8.12 |
| Manylinux2014 (based on Centos 7) | 3.18.3 |
| Manylinux2_24 | 3.22.1 |
| CentOS 7 | 2.8.12 |
| CentOS 8,9 | 3.20.2 |
| openSUSE Tumbleweed | 3.22.1 |
| openSUSE Leap 15 | 3.17.0 |
| RedHat EL 9 (released in November 2021, based of Fedora 34)  | 3.20.5 |
| RedHat EL 8 (released in May 2019, based of Fedora 28)  | 3.14.4 |
| Windows (Chocolatey) | 3.22.1 |
| MacOSX (Homebrew) | 3.22.1 |
| MacOSX (MacPort) | 3.22.1 |
| Gentoo | 3.20.5 |

**Comment by jcfr** [comment]
Based on the last survey, will be moving forward with using the latest patch release of the `3.13` series.
